keps/sig-node/20191205-container-streaming-requests.md

```diff
@@ -0,0 +1,211 @@
+graduated or removed.
+
+[Container Runtime Interface]: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node/container-runtime-interface-v1.md
+[potential SSRF attacks]: https://github.com/kubernetes/kubernetes/issues/85867
+
+### Goals
+
+1. Simplify secure container runtime setup and operation
+2. Remove conformance dependencies on beta features
+
+### Non-Goals
+
+Design a high-bandwidth low-latency production ready streaming channel to containers. Such a system
+should go through an ingress point other than the apiserver, and could be implemented (or at least
+prototyped) out of tree by leveraging the new [ephemeral containers
+feature](20190212-ephemeral-containers.md).
+
+## Proposal
+
+After the [transition period](#rollout-plan), the only way to configure container streaming will be
+using the local redirect with Kubelet proxy (`--redirect-container-streaming=false`) approach. Under
+this configuration, handling streaming redirects are not needed by the apiserver, so
+`StreamingProxyRedirects` will be deprecated and eventually removed. Once `StreamingProxyRedirects`
+are removed, `ValidateProxyRedirects` is no longer needed and can be removed. Finally, the
+`--redirect-container-streaming` flag will be deprecated and eventually removed.
+
+### Rollout Plan
+
+1. release **1.18**
+   1. Mark the `--redirect-container-streaming` flag as deprecated. Log a warning on use.
+   2. Mark the `StreamingProxyRedirects` feature as deprecated. Log a warning on use.
+2. release **1.19** - no changes
+3. release **1.20**
+   1. `--redirect-container-streaming` can no longer be enabled. If the flag is set, log an error
+      and continue as though it is unset.
+4. release **1.21** _(extra safe alternative: 1.22)_
+   1. Default `StreamingProxyRedirects` to disabled. If there is a >= 2 version skew between master
+      and nodes, and the old nodes were enabling `--redirect-container-streaming`, this **will break
+      them**. In this case, the `StreamingProxyRedirects` can still be manually enabled.
+5. release **1.22** _(extra safe alternative: 1.24)_
+   1. Remove the `--redirect-container-streaming` flag.
+   2. Remove the `StreamingProxyRedirects` feature.
+
+### Risks and Mitigations
+
+#### Dependence on apiserver redirects
+
+Risk: a user depends on the ability to handle streaming redirects in the apiserver.
+
+This could be the case if:
+
+1. Pods are run outside the node, and the kubelet has no connectivity to streaming server. This is
+   not a conformant setup, and I do not think we need to support it. If the kubelet _does_ have
+   connectivity, then the kubelet can still proxy the stream.
+2. The use case is extremely latency or resource sensitive, and cannot tolerate the extra hop
+   through the Kubelet. My hypothesis is that these use cases would also be intolerant of the
+   apiserver hop, and should look into alternative solutions.
```

Comment:
The canonical case today is someone doing a database backup via exec.  People do it.  Most of them are doing it through the kubelet.  It can break the Kubelet.  People are roughly ok with it.

Reply:
Bleh, I just realized we aren't putting any limits on the connection. I added a paragraph about introducing a parameter to limit the throughput, a kubelet analog to the apiserver's `MaxConnectionBytesPerSec`